asserts/hardwareidentity.go

```diff
@@ -0,0 +1,147 @@
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+
+package asserts
+
+import (
+	"encoding/base64"
+	"errors"
+	"fmt"
+	"strings"
+
+	"golang.org/x/crypto/sha3"
+)
+
+type HardwareIdentity struct {
+	assertionBase
+
+	issuerId              string
+	manufacturer          string
+	hardwareName          string
+	hardwareId            string
+	hardwareIdKey         string
+	hardwareIdKeySha3384  string
+}
+
+func (h *HardwareIdentity) IssuerId() string {
+	return h.issuerId
+}
+
+func (h *HardwareIdentity) Manufacturer() string {
+	return h.manufacturer
+}
+
+func (h *HardwareIdentity) HardwareName() string {
+	return h.hardwareName
+}
+
+func (h *HardwareIdentity) HardwareId() string {
+	return h.hardwareId
+}
+
+func (h *HardwareIdentity) HardwareIdKey() string {
+	return h.hardwareIdKey
+}
+
+func (h *HardwareIdentity) HardwareIdKeySha3384() string {
+	return h.hardwareIdKeySha3384
+}
+
+func assembleHardwareIdentity(assert assertionBase) (Assertion, error) {
+	issuerId, err := checkNotEmptyString(assert.headers, "issuer-id")         
+	if err != nil {
+		return nil, err
+	}
+	if !validAccountID.MatchString(issuerId) {
+		return nil, fmt.Errorf("invalid issuer id: %s", issuerId)
+	}
+
+	if issuerId != assert.headers["authority-id"] {
+		return nil, errors.New("issuer id must match authority id")
+	}
+
+	manufacturer, err := checkNotEmptyString(assert.headers, "manufacturer")         
+	if err != nil {
+		return nil, err
+	}
+	
+	hardwareName, err := checkStringMatches(assert.headers, "hardware-name", validModel)         
+	if err != nil {
+		return nil, err
+	}
+	
+	hardwareId, err := checkNotEmptyString(assert.headers, "hardware-id")           
+	if err != nil {
+		return nil, err
+	}
+	
+	hardwareIdKey, err := checkNotEmptyString(assert.headers, "hardware-id-key")        
+	if err != nil {
+		return nil, err
+	}
+	
+	if !IsParsablePemPublicKeyBody(hardwareIdKey) {
+		return nil, errors.New(`"hardware-id-key" header should be the body of a PEM`)
+	}
+
+	// don't check if non-empty as check was done for primary key
+	hardwareIdKeySha3384, _ := checkNotEmptyString(assert.headers, "hardware-id-key-sha3-384") 
+
+	
+    // Compute hash and display as hexadecimal
+	hash := sha3.New384()
+    hash.Write([]byte(hardwareIdKey))
+    hashedHardwareIdKey := hash.Sum(nil)
+
+	if hardwareIdKeySha3384 != fmt.Sprintf("%x", hashedHardwareIdKey) {
+		return nil, fmt.Errorf("hardware id key does not match provided sha3 digest")
+	}
+
+	if len(assert.body) != 0 {
+		return nil, errors.New("body must be empty")
+	}
+
+	return &HardwareIdentity{
+		assertionBase: assert,
+		issuerId: issuerId,
+		manufacturer: manufacturer,
+		hardwareName: hardwareName,
+		hardwareId: hardwareId,
+		hardwareIdKey: hardwareIdKey,
+		hardwareIdKeySha3384: hardwareIdKeySha3384,
+	}, nil
+}
+
+
+// IsParsablePemPublicKeyBody checks if s is a valid PEM body for RFC 7468 "PUBLIC KEY"
+func IsParsablePemPublicKeyBody(s string) bool {
```

Comment:
Check if it's easier to use `encoding/pem` instead: https://pkg.go.dev/encoding/pem#example-Decode